Automattic/woocommerce-payments
# Add unit tests for WooPay cart eligibility

#### Changes proposed in this Pull Request





This PR adds additional unit tests around the `WooPay_Utilities:: should_enable_woopay_on_cart_or_checkout` function. Since we had some regressions previously around this area I wanted to make sure we got full coverage of it's functionality.  https://github.com/Automattic/woocommerce-payments/pull/6509 added the key test in checking that the JS config shows that WooPay is properly enabled based on the result of `should_enable_woopay_on_cart_or_checkout` and this PR adds tests that make sure `should_enable_woopay_on_cart_or_checkout` is matching the desired conditions.

The testing of this function was a bit odd as it relied on four key things:
- Which page was being viewed
- Whether the user is logged in
- Whether there is a subscription product in the cart
- Whether guest checkout is enabled

To test these states I used:
- The `woocommerce_is_checkout` filter
- `wp_set_current_user`
- `WC_Subscriptions_Cart::set_cart_contains_subscription`
- `update_option( 'woocommerce_enable_guest_checkout', 'no' );`

This meant I needed to do some clean up after every test which I put into a single private function to reset these states after any relevant tests. **I describe all this here in case there are better ways to test those states and/or better ways to perform that clean up** (note that it doesn't need to happen after every test in the file so I didn't use `tearDown`, but maybe I should?).





#### Testing instructions




1. Confirm that the unit tests pass as they are.
2. Modify something in the code that would change the conditions and confirm that the appropriate unit test fails.

-------------------


**Post merge**
